Zephyr module variables in Kconfig (#27740)
zephyrproject-rtos/zephyr: doc/guides/modules.rst

     cmake: .
     kconfig: Kconfig

Build system integration
========================

When a module has been described using the :file:`module.yml` file then it will
automatically be included into the Zephyr Kconfig and CMake build system.

> To facilitate referencing between Zephyr and Zephyr modules, then all modules

Reviewer: minor edit suggestion, drop the `then`.

Author: fixed

will be accessible through Kconfig and CMake variables.

> In Kconfig and CMake, the ``ZEPHYR_<module-name>_MODULE_DIR`` can be used to

Reviewer 2: I didn't raise a bug about this because it wasn't officially documented before, but this is not really correct.

The existence of a module.yml file does not guarantee this variable is defined, even in CMake.

See this source code comment in one of my PRs for details: https://github.com/zephyrproject-rtos/zephyr/pull/27688/files#diff-a1f9e35746e0274b9ad8cfb12162414dR38

I think this situation should be fixed, or at least documented more clearly.

Author: Agreed, as minimum this should be rephrased to be more precise that only if build files are present in the module.

Guess the introduction of sanity in `module.yml` made a small hole here.

Reviewer 2: > as minimum this should be rephrased to be more precise that only if build files are present in the module.

I agree that would be the minimum acceptable change, since it makes the documentation correct.

Reviewer 2: (Although FTR I would appreciate it if the existence of a module.yml resulted in the definition of ZEPHYR_FOO_MODULE_DIR. That is the best possible world but I agree it's outside the scope of this PR)

Author: couldn't leave code like that.
Fixed. Now docs and code is aligned.

Reviewer 2: OK, rebased and tested on top of the updated version. I confirm the fix. Thanks for fixing :+1:

     cmake: .
     kconfig: Kconfig

Build system integration
========================

When a module has been described using the :file:`module.yml` file then it will
automatically be included into the Zephyr Kconfig and CMake build system.

To facilitate referencing between Zephyr and Zephyr modules all modules
will be accessible through Kconfig and CMake variables.

In Kconfig and CMake, the ``ZEPHYR_<module-name>_MODULE_DIR`` can be used to
obtain the absolute path to the module directory.

In CMake, the ``ZEPHYR_<module-name>_CMAKE_DIR`` can be used to obtain the
absolute path to the :file:`CMakeLists.txt` that is included into CMake build

> system.

Author: Not completely accurate. The  `ZEPHYR_<module-name>_CMAKE_DIR` is only dependent on CMakeLists.txt file.

To use the full path to Zephyr module ``FOO`` use:

- Module directory in CMake: ``${ZEPHYR_FOO_MODULE_DIR}``
- Module directory in Kconfig: ``$(ZEPHYR_FOO_MODULE_DIR)``
- :file:`CMakeLists.txt` directory include in build system: ``${ZEPHYR_FOO_CMAKE_DIR}``

> Those variables may be used to obtain a full path to Zephyr module or to test
> whether a given module is present in the system.
> 
> For example, to verify that ``FOO`` exists in the system and is a Zephyr module
> then one may do:

Reviewer 2: ```suggestion
Notice how a lowercase module name ``foo`` is capitalized to ``FOO``
in both CMake and Kconfig.

These variables can also be used to test whether a given module exists.
For example, to verify that ``foo`` is the name of a Zephyr module:
```

.. code-block:: cmake

  if(ZEPHYR_FOO_MODULE_DIR)
    # Do something if FOO exists.
  endif()

> In Kconfig, the variable may be used to specify a path to include.
> An example can be inclusion of additional Kconfig SoC files.
> 
> In this example ``FOO`` specifies an additional ``SOC_ROOT`` and in
> :file:`<FOO>/soc/arm/foo_soc/Kconfig.soc` one may do:

Reviewer 2: ```suggestion
In Kconfig, the variable may be used to find additional files to include.
For example, to include the file :file:`some/Kconfig` in module ``foo``:
```

.. code-block:: kconfig

>   source "$(ZEPHYR_FOO_MODULE_DIR)/soc/arm/foo_soc/Kconfig.soc.common"

Reviewer 2: ```suggestion
  source "$(ZEPHYR_FOO_MODULE_DIR)/some/Kconfig"
```

During CMake processing of each Zephyr module, the following two variables are
also available:

> - Module directory in CMake of current module being processed: ``${ZEPHYR_CURRENT_MODULE_DIR}``
> - :file:`CMakeLists.txt` directory of current module directory: ``${ZEPHYR_CURRENT_CMAKE_DIR}``

Reviewer 2: ```suggestion
- the current module's top level directory: ``${ZEPHYR_CURRENT_MODULE_DIR}``
- the current module's :file:`CMakeLists.txt` directory: ``${ZEPHYR_CURRENT_CMAKE_DIR}``
```

> This remove the need for a Zephyr module to know its own name during CMake
> processing, and the module can source additional CMake files using the current
> variable, for example:

Reviewer 2: ```suggestion
This removes the need for a Zephyr module to know its own name during CMake
processing. The module can source additional CMake files using these ``CURRENT``
variables. For example:
```

Reviewer 2: @tejlmand it occurs to me that this means no module can be named "current".

Worth enforcing in the build system?

Author: Good catch. Let's hope someone doesn't create a current measurement sw tool, and call it `current`.

Author: added extra commit to check for `current`: https://github.com/zephyrproject-rtos/zephyr/pull/27740/commits/8511f9fd9a854c1963a415a6a5e93681eaf0d79b

Reviewer 2: > added extra commit to check for `current`: [8511f9f](https://github.com/zephyrproject-rtos/zephyr/commit/8511f9fd9a854c1963a415a6a5e93681eaf0d79b)

nice :+1:
